sdk/ai/ai-projects/src/tracing.ts

```diff
@@ -0,0 +1,278 @@
+    GEN_AI_SYSTEM_MESSAGE = "gen_ai.system.message",
+    GEN_AI_EVENT_CONTENT = "gen_ai.event.content",
+    ERROR_TYPE = "error.type"
+}
+export enum TrackingOperationName {
+    CREATE_AGENT = "create_agent",
+    CREATE_THREAD = "create_thread",
+    CREATE_MESSAGE = "create_message",
+    CREATE_RUN = "create_run",
+    START_THREAD_RUN = "start_thread_run",
+    EXECUTE_TOOL = "execute_tool",
+    LIST_MESSAGES = "list_messages",
+    SUBMIT_TOOL_OUTPUTS = "submit_tool_outputs",
+    CREATE_THREAD_RUN = "create_thread_run",
+}
+
+export interface TracingAttributeOptions {
+    operationName?: string;
+    name?: string | null;
+    description?: string | null;
+    serverAddress?: string | null;
+    threadId?: string | null;
+    agentId?: string | null;
+    instructions?: string | null;
+    additional_instructions?: string | null;
+    runId?: string | null;
+    runStatus?: string | null;
+    responseModel?: string | null;
+    model?: string | null;
+    temperature?: number | null;
+    topP?: number | null;
+    maxPromptTokens?: number | null;
+    maxCompletionTokens?: number | null;
+    responseFormat?: string | null;
+    genAiSystem?: string | null;
+    messageId?: string | null;
+    messageStatus?: string | null;
+    eventContent?: string | null;
+    usagePromptTokens?: number | null;
+    usageCompletionTokens?: number | null;
+}
+
+export class TracingUtility {
+    private static tracingClient = createTracingClient({
+        namespace: "Microsoft.CognitiveServices",
+        packageName: "@azure/ai-projects",
+        packageVersion: "1.0.0-beta.1",
+
+    });
+
+    static async withSpan<Options extends { tracingOptions?: OperationTracingOptions },
+        Request extends (updatedOptions: Options) => ReturnType<Request>>(name: string, options: Options, request: Request,
+            startTrace?: (span: Omit<TracingSpan, "end">, updatedOptions: Options,) => void,
+            endTrace?: (span: Omit<TracingSpan, "end">, updatedOptions: Options, result: ReturnType<Request>) => void,
+        ): Promise<Resolved<ReturnType<Request>>> {
+        return TracingUtility.tracingClient.withSpan(name, options, async (updatedOptions: Options, span: Omit<TracingSpan, "end">) => {
+            if (startTrace) {
+                try {
+                    startTrace(span, updatedOptions);
+                }
+                catch { /* empty */ }
+
+            }
+            let result: ReturnType<Request> | undefined;
+            try {
+                result = await request(updatedOptions);
+            } catch (error) {
+                const errorStatus: SpanStatusError = { status: "error" }
+                if (error instanceof Error) {
+                    errorStatus.error = error;
+                }
+                throw error;
+            }
+
+            if (endTrace && result !== undefined) {
+                try {
+                    endTrace(span, updatedOptions, result);
+                } catch { /* empty */ }
```

Comment:
Consider logging the error so the user can understand why ending the trace did not happen